### PR TITLE
Pod GC controller - use node lister

### PR DESCRIPTION
**Pull request by jkaniuk**
Change-Id: Ie5e97dac8e434fa032af2a2923dd66df63b4286c

**What type of PR is this?**
/kind cleanup
/sig scalability

**What this PR does / why we need it**:
Pod GC controller lists the nodes in the cluster in every loop, which is a costly operation in a big cluster.
This PR changes controller implementation to use node informer events, only occasionally listing all the nodes if watch is stale.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

### CODE REVIEW

File: pkg/controller/podgc/gc_controller.go [hunk 1]
```diff
@@ -185,6 +186,114 @@ func (gcc *PodGCController) gcUnscheduledTerminating(pods []*v1.Pod) {
...
+}
+
+func (dnw *deletedNodesWatcher) deleteNode(obj interface{}) {
+	node := obj.(*v1.Node)
+	dnw.existingNodes.Delete(node.Name)
+	dnw.markNodeDeleted(node.Name)
+}
+
+func (dnw *deletedNodesWatcher) markNodeDeleted(name string) {
+	dnw.recentlyDeletedNodes[name] = time.Now()
```
Comment (caesarxuchao) [inline pkg/controller/podgc/gc_controller.go]: Should the recentlyDeletedNodes and existingNodes be protected by locks?

File: pkg/controller/podgc/gc_controller.go [hunk 2]
```diff
@@ -185,6 +186,114 @@ func (gcc *PodGCController) gcUnscheduledTerminating(pods []*v1.Pod) {
...
+		if _, found := dnw.recentlyDeletedNodes[name]; found {
+			deletedNames.Insert(name)
+			continue
+		}
+		if _, found := dnw.existingNodes[name]; found {
+			continue
+		}
+		unknownNames.Insert(name)
+	}
+	if len(unknownNames) > 0 {
```
Comment (caesarxuchao) [inline pkg/controller/podgc/gc_controller.go]: If a user keeps creating pods with fake node names, than updateExistingNodes() will always be called.

I'm thinking if there is a better way to detect if the cached node list is stale.

File: pkg/controller/podgc/gc_controller.go [hunk 3]
```diff
@@ -38,6 +38,8 @@ import (
 
 const (
 	gcCheckPeriod = 20 * time.Second
+	purgePeriod   = 20 * time.Minute
```
Comment (mm4tt) [inline pkg/controller/podgc/gc_controller.go]: Can we add a documentation for these constants?

File: pkg/controller/podgc/gc_controller.go [hunk 4]
```diff
@@ -65,6 +70,10 @@ func NewPodGC(kubeClient clientset.Interface, podInformer coreinformers.PodInfor
 
 	gcc.podLister = podInformer.Lister()
 	gcc.podListerSynced = podInformer.Informer().HasSynced
+	gcc.podListerSynced = podInformer.Informer().HasSynced
```
Comment (mm4tt) [inline pkg/controller/podgc/gc_controller.go]: Duplicated line?

File: pkg/controller/podgc/gc_controller.go [hunk 5]
```diff
@@ -122,40 +131,32 @@ func (gcc *PodGCController) gcTerminated(pods []*v1.Pod) {
...
-			}
-		}(terminatedPods[i].Namespace, terminatedPods[i].Name)
-	}
-	wait.Wait()
 }
 
 // gcOrphaned deletes pods that are bound to nodes that don't exist.
 func (gcc *PodGCController) gcOrphaned(pods []*v1.Pod) {
 	klog.V(4).Infof("GC'ing orphaned")
-	// We want to get list of Nodes from the etcd, to make sure that it's as fresh as possible.
```
Comment (mm4tt) [inline pkg/controller/podgc/gc_controller.go]: QQ, now that we have the watch on nodes, cannot we change the approach and instead of calling this method periodically call this method every time we get a "node deleted" event to see if there aren't any pods that became orphaned? I might be wrong, but that should guarantee eventual consistency and would be simpler than what you have in the FileterDeletedNodeNames, WDYT?

Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: Implementation could probably be changed to use only watches. I am not sure if that would be substantially simpler. Regarding your exact question:
1. We need to handle restarting controller, when there is no access to old deleted events
2. Do we care about a time between pod becomes orphaned and its removal? Both in case of node being deleted and pod assigned to a not-ever-existing node.
3. Other methods in this controller run periodically. Is there any benefit of running this method only on node deletion (assuming using only data from watches)?

File: pkg/controller/podgc/gc_controller.go (hunk 4 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: fixed

File: pkg/controller/podgc/gc_controller.go [hunk 6]
```diff
@@ -185,6 +186,114 @@ func (gcc *PodGCController) gcUnscheduledTerminating(pods []*v1.Pod) {
...
+		return err
+	}
+	for i := range nodesList.Items {
+		dnw.existingNodes.Insert(nodesList.Items[i].Name)
+	}
+	return nil
+}
+
+func (dnw *deletedNodesWatcher) purgeIfNeeded() {
+	if !time.Now().Add(-purgePeriod).After(dnw.lastPurge) {
```
[the remark this answers is not in the document]
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: fixed

File: pkg/controller/podgc/gc_controller.go (hunk 1 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: Right, added guards

File: pkg/controller/podgc/gc_controller.go (hunk 3 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: Done

File: pkg/controller/podgc/gc_controller.go (hunk 2 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: I've changed the implementation. Currently in fake node names scenario, there is a single GET (per ~hour) for every unique node name.

**Review by mm4tt — COMMENTED** [review]
As discussed in person, it would be worth to consider a simpler solution for gcOrphaned. 
The way I see it would be

1. In gcOrphaned method, we take list of pods and list of nodes from informers and do a diff to find potential deleted nodes
2. For each potential deleted node we add its node name to a workqueue using AddAfter(nodeName, 30s) to address the case where node informer is lagging behind pod informer
3. Now, a worker processing items from the above workqueue would do the following
  3.1 Check via informer if node appeared
  3.2 If not, issue a GET asking for this node (to address lagging informer)
  3.3 If node not found then get all pods assigned to this node from podInformer and delete them

This should be much simpler (I'd say 30 lines of code vs over 200) and easier to analyze in terms of potential race conditions. @wojtek-t, thoughts?

**Comment by wojtek-t** [comment]
> This should be much simpler (I'd say 30 lines of code vs over 200) and easier to analyze in terms of potential race conditions. @wojtek-t, thoughts?

I would need to see the end code for it, but I agree it sounds promising.

File: pkg/controller/podgc/gc_controller.go [hunk 7]
```diff
@@ -141,23 +157,36 @@ func (gcc *PodGCController) gcTerminated(pods []*v1.Pod) {
...
+	deletedNodesNames := sets.NewString()
+	for _, nodeName := range gcc.nodeQueue.GetWork() {
+		if existingNodeNames.Has(nodeName) {
 			continue
 		}
-		if nodeNames.Has(pod.Spec.NodeName) {
+		exists, err := gcc.checkIfNodeExists(nodeName)
+		if err != nil {
+			klog.Errorf("Error while getting node %q: %v", nodeName, err)
+			continue
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller.go]: Error handling is broken here - if we couldn't get node due to some transient error, we should requeue it retry or sth like that.

File: pkg/controller/podgc/queue.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+// UniqueWorkQueue allows queuing items with a timestamp. An item is
+// considered ready to process if the timestamp has expired.
+type UniqueWorkQueue interface {
```
Comment (wojtek-t) [inline pkg/controller/podgc/queue.go]: I'm generally not a fan of creating new structures. 
I think we can simply use DelayingQueue:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L37

Basically it gives you "Get()" and "AddAfter()" methods, which seem to be all you need.

File: pkg/controller/podgc/gc_controller_test.go [hunk 9]
```diff
@@ -135,90 +150,203 @@ func TestGCTerminated(t *testing.T) {
...
+			}
+			creationTime := time.Unix(0, 0)
+			for _, pod := range test.pods {
+				creationTime = creationTime.Add(1 * time.Hour)
+				podInformer.Informer().GetStore().Add(&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{Name: pod.name, CreationTimestamp: metav1.Time{Time: creationTime}},
+					Status:     v1.PodStatus{Phase: pod.phase},
+					Spec:       v1.PodSpec{NodeName: pod.nodeName},
+				})
+			}
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller_test.go]: I think the test would be simpler if you would test gcc.gcOrphaned (and the value would be the same).
What's the value of adding stuff to informers?

File: pkg/controller/podgc/gc_controller.go [hunk 10]
```diff
@@ -37,35 +39,45 @@ import (
 )
 
 const (
+	// gcCheckPeriod defines frequency of running main controller loop
 	gcCheckPeriod = 20 * time.Second
+	// quarantineTime defines how long unknown nodes wait before check
+	quarantineTime = 30 * time.Second
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller.go]: It seems we're running GC every 20s. Shouldn't we keep those two values in sync? (maybe actually it should be the same value?)

File: pkg/controller/podgc/gc_controller_test.go [hunk 11]
```diff
@@ -135,90 +150,203 @@ func TestGCTerminated(t *testing.T) {
...
+				}
+			}
+			for _, change := range test.clientChanges {
+				node := testutil.NewNode(change.name)
+				switch change.kind {
+				case watch.Added:
+					client.CoreV1().Nodes().Create(node)
+				case watch.Deleted:
+					client.CoreV1().Nodes().Delete(change.name, &metav1.DeleteOptions{})
+				}
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller_test.go]: Yeah - the more I look into this test, the more I'm convinced we should modify informers.
I think you should just update client, then List from client and pass the list of pods and nodes to gcOrphaned.

File: pkg/controller/podgc/gc_controller_test.go [hunk 12]
```diff
@@ -135,90 +150,203 @@ func TestGCTerminated(t *testing.T) {
...
+				i, test.deletedPodNames.List(), deletedPodNames)
 		}
 	}
 }
 
 func TestGCOrphaned(t *testing.T) {
-	type nameToPhase struct {
-		name  string
-		phase v1.PodPhase
+	type podMeta struct {
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller_test.go]: i don't see the value of having podMeta type here, and first creating podmeta and then creating Pod from it.
Instead, just:
- create a makePod function (sth like:
```
makePod := func(name, nodeName string, phase v1.PodPhase) *v1.Pod {
  return &v1.Pod{
    ObjectMeta: metav1.ObjectMeta{Name: name},
    Spec:       v1.PodSpec{NodeName: nodeName},
    Status:     v1.PodStatus{Phase: phase},
  }
}
```
and use it below (i.e. change pods to be of type []*v1.Pod

File: pkg/controller/podgc/gc_controller_test.go [hunk 13]
```diff
@@ -135,90 +150,203 @@ func TestGCTerminated(t *testing.T) {
...
+		name     string
+		nodeName string
+		phase    v1.PodPhase
+	}
+	type nodeEvent struct {
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller_test.go]: I guess, same here. Why can't we have explicit fields:
```
nodesToAdd []*v1.Node
nodesToDelete []*v1.Node
```

**Review by wojtek-t — COMMENTED** [review]
I added couple comments, but at high-level I really like this approach.

File: pkg/controller/podgc/gc_controller.go (hunk 7 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: it will be requeued like any new node

File: pkg/controller/podgc/queue.go (hunk 8 above)
Comment (jkaniuk) [reply pkg/controller/podgc/queue.go]: DelayingQueue does not have ability to deduplicate items.

File: pkg/controller/podgc/gc_controller_test.go (hunk 9 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller_test.go]: gcOrhpaned depends on the up to date set of nodes and (possibly stale) state of lister.
I'm testing different combinations of both values before and after quarantine.
It could be done by passing lister state to gcOrhpaned directly, but that would still need 3-4 sets of nodes in tests.

File: pkg/controller/podgc/gc_controller.go (hunk 10 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: I think those values should be separated.
Selected value for both could be currently in sync, but do not have to be later (e.g. when 30-40s turns out to be too less under load).

File: pkg/controller/podgc/gc_controller_test.go (hunk 11 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller_test.go]: I am explicitly testing scenarios when client and informers are not in sync

File: pkg/controller/podgc/gc_controller_test.go (hunk 13 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller_test.go]: We could, but the beauty of that is deceptive.
nodesToAddToClient, nodesToDeleteFromClient, nodesToAddToInformer, nodesToDeleteFromInformer?

File: pkg/controller/podgc/queue.go (hunk 8 above)
Comment (wojtek-t) [reply pkg/controller/podgc/queue.go]: IIRC - it's impossible to have mutliple items in Queue (and thus DelayingQueue) with the same key at the same time.
So it's kind-of deduping out-of-the-box.

File: pkg/controller/podgc/gc_controller.go (hunk 7 above)
Comment (wojtek-t) [reply pkg/controller/podgc/gc_controller.go]: How?

File: pkg/controller/podgc/gc_controller_test.go (hunk 13 above)
Comment (wojtek-t) [reply pkg/controller/podgc/gc_controller_test.go]: You kind of already has it - you're just hiding this in the "event.Type". being more explicit and avoiding switches below sounds like a win to me.
Also, you're not handling e.g. update event type, so that will at least make it explicit.

File: pkg/controller/podgc/gc_controller_test.go (hunk 12 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller_test.go]: OK

File: pkg/controller/podgc/gc_controller_test.go (hunk 13 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller_test.go]: OK

File: pkg/controller/podgc/gc_controller.go (hunk 7 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: There would be still orphaned pods so their node would end up in the queue again.

File: pkg/controller/podgc/gc_controller.go [hunk 14]
```diff
@@ -78,6 +93,9 @@ func (gcc *PodGCController) Run(stop <-chan struct{}) {
 	if !cache.WaitForNamedCacheSync("GC", stop, gcc.podListerSynced) {
 		return
 	}
+	if !cache.WaitForNamedCacheSync("GC", stop, gcc.nodeListerSynced) {
```
[the remark this answers is not in the document]
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: Done

File: pkg/controller/podgc/gc_controller.go (hunk 7 above)
Comment (wojtek-t) [reply pkg/controller/podgc/gc_controller.go]: I need to think about it - but assuming you're right, it at least requires a comment why you're not taking any action on error here.

File: pkg/controller/podgc/gc_controller_test.go (hunk 9 above)
Comment (wojtek-t) [reply pkg/controller/podgc/gc_controller_test.go]: OK fair enough - that sounds reasonable to me.

File: pkg/controller/podgc/queue.go (hunk 8 above)
Comment (mm4tt) [reply pkg/controller/podgc/queue.go]: +1, could you rewrite to DelayingQueue?
As Wojtek said, it dedupes items, there is even a test for that - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/util/workqueue/delaying_queue_test.go#L72

File: pkg/controller/podgc/gc_controller.go (hunk 7 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: Done

File: pkg/controller/podgc/queue.go (hunk 8 above)
Comment (jkaniuk) [reply pkg/controller/podgc/queue.go]: Done

File: pkg/controller/podgc/gc_controller_test.go (hunk 12 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller_test.go]: Done

File: pkg/controller/podgc/gc_controller.go (hunk 10 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: Changed to 40s to be less confusing.

File: pkg/controller/podgc/gc_controller.go [hunk 15]
```diff
@@ -75,25 +87,31 @@ func (gcc *PodGCController) Run(stop <-chan struct{}) {
...
 
-	if !cache.WaitForNamedCacheSync("GC", stop, gcc.podListerSynced) {
+	if !cache.WaitForNamedCacheSync("GC", stop, gcc.podListerSynced, gcc.nodeListerSynced) {
 		return
 	}
 
 	go wait.Until(gcc.gc, gcCheckPeriod, stop)
 
 	<-stop
+	gcc.nodeQueue.ShutDown()
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller.go]: Move it to defer (right above defer klog.Infof...)

File: pkg/controller/podgc/gc_controller.go [hunk 16]
```diff
@@ -141,23 +158,44 @@ func (gcc *PodGCController) gcTerminated(pods []*v1.Pod) {
...
+		if pod.Spec.NodeName != "" && !existingNodeNames.Has(pod.Spec.NodeName) {
+			gcc.nodeQueue.AddAfter(pod.Spec.NodeName, quarantineTime)
+		}
+	}
+	// Check if nodes are still missing after quarantine period
+	deletedNodesNames := sets.NewString()
+	for gcc.nodeQueue.Len() > 0 {
+		item, quit := gcc.nodeQueue.Get()
+		if quit {
+			klog.Errorf("Node queue closed, exiting")
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller.go]: This isn't an error - simply remove this log.

File: pkg/controller/podgc/gc_controller.go [hunk 17]
```diff
@@ -141,23 +158,44 @@ func (gcc *PodGCController) gcTerminated(pods []*v1.Pod) {
...
+			return
+		}
+		gcc.nodeQueue.Done(item)
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller.go]: Done() should be called after the item is fully processed.
So probably the easiest way to do that is to opaque it with a function and use defer.

Comment (wojtek-t) [reply pkg/controller/podgc/gc_controller.go]: It would be good to split this function out for readability too.

File: pkg/controller/podgc/gc_controller.go [hunk 18]
```diff
@@ -169,6 +207,17 @@ func (gcc *PodGCController) gcOrphaned(pods []*v1.Pod) {
...
 
+func (gcc *PodGCController) checkIfNodeExists(name string) (bool, error) {
+	_, fetchErr := gcc.kubeClient.CoreV1().Nodes().Get(name, metav1.GetOptions{})
+	if fetchErr == nil {
+		return true, nil
+	}
+	if errors.IsNotFound(fetchErr) {
+		return false, nil
+	}
+	return false, fetchErr
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller.go]: Why not simply:
```
if errors.IsNotFound(fetchErr) {
  return false, nil
}
return fetchErr == nil, fetchErr
```

File: pkg/controller/podgc/gc_controller_test.go [hunk 19]
```diff
@@ -47,12 +50,25 @@ func (*FakeController) LastSyncResourceVersion() string {
...
-	controller := NewPodGC(kubeClient, podInformer, terminatedPodThreshold)
+	nodeInformer := informerFactory.Core().V1().Nodes()
+	controller := NewPodGC(kubeClient, podInformer, nodeInformer, terminatedPodThreshold)
 	controller.podListerSynced = alwaysReady
-	return controller, podInformer
+	return controller, podInformer, nodeInformer
+}
+
+func compareStringSetToList(set sets.String, list []string) bool {
+	for _, pod := range list {
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller_test.go]: s/pod/item/ (function is generic enough)

File: pkg/controller/podgc/gc_controller_test.go [hunk 20]
```diff
@@ -135,90 +151,230 @@ func TestGCTerminated(t *testing.T) {
...
-	type nameToPhase struct {
-		name  string
-		phase v1.PodPhase
+func makePod(name string, nodeName string, phase v1.PodPhase) *v1.Pod {
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Status: v1.PodStatus{Phase: phase},
+		Spec:   v1.PodSpec{NodeName: nodeName},
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller_test.go]: nit: move Spec above added

File: pkg/controller/podgc/gc_controller_test.go [hunk 21]
```diff
@@ -135,90 +151,230 @@ func TestGCTerminated(t *testing.T) {
...
+			delay:              2 * quarantineTime,
+			addedInformerNodes: []*v1.Node{testutil.NewNode("node")},
+			pods: []*v1.Pod{
+				makePod("a", "node", v1.PodFailed),
+			},
+			itemsInQueue:    1,
+			deletedPodNames: sets.NewString(),
+		},
+		{
+			name:               "node deleted from client after quarantine",
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller_test.go]: Not sure I understand this case - client will never lag comparing to informer.

File: staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go [hunk 22]
```diff
@@ -35,14 +35,17 @@ type DelayingInterface interface {
 
 // NewDelayingQueue constructs a new workqueue with delayed queuing ability
 func NewDelayingQueue() DelayingInterface {
-	return newDelayingQueue(clock.RealClock{}, "")
+	return NewDelayingQueueWithCustomClock(clock.RealClock{}, "")
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go]: Please move the client-go changes to a separate PR.

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml [hunk 23]
```diff
@@ -872,6 +872,7 @@ items:
     resources:
     - nodes
     verbs:
+    - get
```
Comment (wojtek-t) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml]: Auth changes should go to a separate commit.

**Review by wojtek-t — COMMENTED** [review]
Mostly minor comments at this point - once addressed LGTM.

File: pkg/controller/podgc/gc_controller.go [hunk 24]
```diff
@@ -31,41 +32,52 @@ import (
...
+	"k8s.io/client-go/util/workqueue"
 	"k8s.io/component-base/metrics/prometheus/ratelimiter"
 
 	"k8s.io/klog"
 )
 
 const (
+	// gcCheckPeriod defines frequency of running main controller loop
 	gcCheckPeriod = 20 * time.Second
+	// quarantineTime defines how long unknown nodes wait before check
```
Comment (mm4tt) [inline pkg/controller/podgc/gc_controller.go]: Please, please be more elaborative here :) Be more precise about what is being checked for unknown nodes.

**Review by mm4tt — COMMENTED** [review]
LGTM assuming you address Wojtek's comments.

I really like this approach!

File: pkg/controller/podgc/gc_controller_test.go (hunk 21 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller_test.go]: This is sanity check

Comment (wojtek-t) [reply pkg/controller/podgc/gc_controller_test.go]: Add a comment then that this is more a sanity-check. I can imagine that at some point this test may start passing and it may be fine to remove it for a good reason.

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml (hunk 23 above)
Comment (jkaniuk) [reply plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml]: Done

File: staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go (hunk 22 above)
Comment (jkaniuk) [reply staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go]: Done

File: pkg/controller/podgc/gc_controller.go (hunk 24 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: Done

File: pkg/controller/podgc/gc_controller_test.go (hunk 21 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller_test.go]: Done

File: pkg/controller/podgc/gc_controller_test.go (hunk 20 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller_test.go]: Done

File: pkg/controller/podgc/gc_controller_test.go (hunk 19 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller_test.go]: Done

File: pkg/controller/podgc/gc_controller.go (hunk 18 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: Done

File: pkg/controller/podgc/gc_controller.go (hunk 17 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: Done

File: pkg/controller/podgc/gc_controller.go (hunk 15 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: Done

File: cmd/kube-controller-manager/app/core.go [hunk 25]
```diff
@@ -347,6 +347,7 @@ func startPodGCController(ctx ControllerContext) (http.Handler, bool, error) {
 	go podgc.NewPodGC(
 		ctx.ClientBuilder.ClientOrDie("pod-garbage-collector"),
 		ctx.InformerFactory.Core().V1().Pods(),
+		ctx.InformerFactory.Core().V1().Nodes(),
```
Comment (wojtek-t) [inline cmd/kube-controller-manager/app/core.go]: This should go do the podgc-related commit

File: pkg/controller/podgc/gc_controller_test.go [hunk 26]
```diff
@@ -135,90 +151,230 @@ func TestGCTerminated(t *testing.T) {
...
+			delay:              2 * quarantineTime,
+			addedInformerNodes: []*v1.Node{testutil.NewNode("node")},
+			pods: []*v1.Pod{
+				makePod("a", "node", v1.PodFailed),
+			},
+			itemsInQueue:    1,
+			deletedPodNames: sets.NewString(),
+		},
+		{
+			name:               "node deleted from client after quarantine (just sanity check)",
```
Comment (wojtek-t) [inline pkg/controller/podgc/gc_controller_test.go]: Please add an explicit comment, sth like:
// It shouldn't happen that client will be lagging behind informer.
// This test case is more a sanity check.

**Review by wojtek-t — COMMENTED** [review]
Just two minor nits - once applied will approve

File: cmd/kube-controller-manager/app/core.go (hunk 25 above)
Comment (jkaniuk) [reply cmd/kube-controller-manager/app/core.go]: Done

File: pkg/controller/podgc/gc_controller_test.go (hunk 26 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller_test.go]: Done

File: pkg/controller/podgc/gc_controller.go (hunk 2 above)
Comment (jkaniuk) [reply pkg/controller/podgc/gc_controller.go]: Implementation changed once again.
When orphaned pod is found and after 40s there is still no node with matching name in an informer, GET call is issued to verify that the node is truly gone (single call for all pods from the same node, but without caching the result).

**Comment by mm4tt** [comment]
/lgtm

Nice!

File: pkg/controller/podgc/gc_controller.go [hunk 27]
```diff
@@ -141,23 +159,26 @@ func (gcc *PodGCController) gcTerminated(pods []*v1.Pod) {
...
 	for _, pod := range pods {
-		if pod.Spec.NodeName == "" {
-			continue
+		if pod.Spec.NodeName != "" && !existingNodeNames.Has(pod.Spec.NodeName) {
+			gcc.nodeQueue.AddAfter(pod.Spec.NodeName, quarantineTime)
 		}
-		if nodeNames.Has(pod.Spec.NodeName) {
+	}
+	// Check if nodes are still missing after quarantine period
+	deletedNodesNames, quit := gcc.discoverDeletedNodes(existingNodeNames)
```
Comment (tedyu) [inline pkg/controller/podgc/gc_controller.go]: If this is lifted ahead of line 168, we can iterate over pods once - checking deletedNodesNames and existingNodeNames
